### DISCUSSION TITLE
What should go in 8.x

### DISCUSSION THREAD

**Discussion by MikeNeilson** [opendcs/opendcs#39] [Ideas]
Hypothetically what do people, internally (not features), want to see done to ease development.

Some I can think of
- Find a way to normalize the DECODES vs CP side of things. With the history of OpenDCS the two systems some parts of the design can be very different.
  -   Likely there would still be two since some systems ONLY need DECODES but sharing a similar design would make dev work easier.
- Some baseline code needs to know what the database is (CWMS,HDB, reference implementation) finding a way to eliminate that would make the design cleaner.
- For comps threading vs multiple copies of compproc
  -  Both are valid, though I think there are some cases were threading would help, like having a real-time vs older data thread.

Anyways none of those are requirements, just thoughts on the future.

**Comment by MikeNeilson** [comment]
@opendcs/opendcs-core-devs 
@opendcs/opendcs-triange 

I've updated the title, I think we likely need some big changes and a place to start them. I'm going to go ahead an start an 8.x branch; mostly like just renaming my enhancement/ivy_multiproject branch. For now most development should happen on master and be part of 7.x, we can merge new things in as time allots and depending on where we are with various tasks below. When I have more time I'll setup all the issues and projects to track the work below.

Some things I plan to do in 8.0:

- multi project build, baseline is already there
  - baseline I have now is "opendcs.jar", "documentation", and "installer"
  - plan is to even split up opendcs into sub-components, like the DAI classes should be in an "opendcs-api"
  - LRGS (that which is shared would go into a "opendcs-core" library
  - decodes
- annotations and an annotation processor for
 - computations/algorithms
 - data sources
 - batch/shell scripts
 - Thing I probably haven't thought of yet
- flywaydb (unless someone objects) for database management
  - This would also come along with the installer so that either the installer or a to-be-made application can handle database updates on behalf of the user.
- improvements to installer for database install
- build tool helper function to standup an developer environment
- automated tests (hopefully will get hat into 7.x/master as well)
- either
  - moving all SQL to our custom new lambda code
  - migrating SQL to use JDBI (https://jdbi.org/, does the same thing but wraps a lot of the differences between databases)
- probably migrating everything to java.util.logging
- targeting Java 11 as its getting hard to find up-to-date libraries in even java 8 now, plus the "var" keyword can be really handy.
- Optional SSL for DDS server/client
- official (from us) docker containers


Milestone for release is modules/jars are reasonably split and annotations are to a place were we don't have to manually curate any xml files or other data that could be reasonably generated.


I'm happy to stick with ant+ivy for the build, having don't so much work with it now I don't see a super compelling reason to change, unless anyone wants to provide a solid argument to move to gradle (a majority just saying yes is a solid argument). I don't see maven working as well for a project like this. NOTE: for izpack if we don't want to go to izpack 5, we can still use ant for just that part and call ant from gradle (both maven and gradle support calling an ant build.)

Some maybes are, but likely 8.1+:
- moving the XML database to SQLite. That would simplify the structure, but still allow a user to manually edit things. It's rare that a system doesn't have the sqlite3 command.
- LXC image or something that can easily be put on a raspberry pi. Or even just a prebuilt debian image.

**Reply by adamkorynta** [reply]
Why not target JDK 17 as the newer LTS release? And JDK 21 is only a year away at this point.

I'd vote for investing in gradle especially when splitting into sub-projects/modules/components. It makes more sense for the tighter integrations with IDE's. 

Also I'd put automated tests at the top of the list (even though I don't think that list is ordered).

**Reply by MikeNeilson** [reply]
The automated test are (except the currently multi project split) step 1, otherwise I'm not going to know what's actually working right.

I'm not opposed to that big of a jump. if no one objects I'll just do it.

**Reply by adamkorynta** [reply]
Changes to the JDK between 11 and what is proposed for 21 are pretty minimal. Having automated tests setup first, we can always run them in environments with 11, 17, and pre-release 21 to ensure compatibility.

**Reply by DanielTOsborne** [reply]
I think the concern I have jumping so far ahead is what JREs we have available on Solaris.

**Reply by adamkorynta** [reply]
Good point, we should probably add the list of supported operating systems in the readme file.

**Reply by MikeNeilson** [reply]
The usace support contract with oracle covers the jdks and jres on solaris. So we shoukd be covered. But i like Adam's idea with the testing. Maybe we just target 11 for now as there are some mahors changes for us byt we make sure to test the others.

That seems to be fairly common still. Just targetting the java 11 source and binary format.

**Reply by DanielTOsborne** [reply]
The jump to 11 is going to be the most work anyway. Once we're on that, it shouldn't matter too much.

**Reply by MikeNeilson** [reply]
> The automated test are (except the currently multi project split) step 1, otherwise I'm not going to know what's actually working right.

... thinking about it, the automated tests might be step 2... step one is being able to actually setup an environment.. likely going to need to tweak the database scripts a little. Though I think I'll just start simple not with a whole flyway migration.
